### PR TITLE
Multi line merge functionality for merge library.

### PR DESCRIPTION
**Pull request by rakesh-verma-16**
No Ready completely. Works fine for changes in same number of lines. 
Adding functionality for Addition and removal of lines as well.
Will take care of keys addition and removal after that.
Tested at my end only. No tests have been added in the test file.

### CODE REVIEW

File: test.php [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: rakesh
+ * Date: 15/6/16
+ * Time: 9:54 AM
+ */
```
Comment (rakesh-verma-16) [inline test.php]: Will remove this file in my final commits, using it for simple testing.

File: src/ThreeWayMerge.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                $merged[$key] = $y;
+            }
+            else {
+                throw new Exception("A conflict has occured");
+            }
+        }
+
+        return $merged[$key];
+    }
+}
```
Comment (rakesh-verma-16) [inline src/ThreeWayMerge.php:330]: Will add comments and follow standard after I have achieved desired functionality.

File: src/ThreeWayMerge.php [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                        throw new Exception("A conflict has occured");
+                    }
+                    if ($counter >= $count) {
+                        break;
+                    }
+                }
+            }
+            $merged[$key] = implode(PHP_EOL, $merged);
+//        }
+//        else {
```
Comment (rakesh-verma-16) [inline src/ThreeWayMerge.php]: Will clean the code soon.

File: src/ThreeWayMerge.php [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        }
+        if(strpos($z, "\n") !== FALSE) {
+            $remote = explode("\n", $z);
+        }
+
+        else{
+            $remote = array($z);
+        }
+//        print_r($ancestor);
+//        if(isset($remote) || isset($ancestor) || isset($local)) {
```
Comment (rakesh-verma-16) [inline src/ThreeWayMerge.php]: Will clean the code in upcoming commits.

File: src/ThreeWayMerge.php [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            if ($ancestor[$key] == $local[$key]) {
+                $merged[$key] = $remote[$key];
+            } elseif ($ancestor[$key] == $remote[$key] || $local[$key] == $remote[$key]) {
+                $merged[$key] = $local[$key];
+            } else {
+                throw new Exception("A conflict has occured");
+            }
+        }
+        for ($i = $count_ancestor; $i < $count; $i++) {
+            $count == $count_local ? $merged[$i] = $local[$i] : $merged[$i] = $remote[$i];
```
Comment (rakesh-verma-16) [inline src/ThreeWayMerge.php]: Will produce bugs, will be fixing it in next commit.

File: src/ThreeWayMerge.php [hunk 6]
```diff
@@ -0,0 +1,206 @@
+<?php
```
Comment (sharprakeshverma) [inline src/ThreeWayMerge.php]: Will add docs soon

File: tests/ThreeWayMergeTest.php [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+        ];
+
+        $remote = [
+            'keyA' => ['This is not fun'],
+        ];
+
+        $expected = [
+            'keyA' => ['This is fun
+            it was easy'],
+        ];
```
Comment (dickolsson) [inline tests/ThreeWayMergeTest.php:364]: Why do we expect `it was easy` here? Too me it seems like both the local and remote versions have removed this line.

File: tests/ThreeWayMergeTest.php [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+        $remote = [
+            'keyA' => ['This is not fun
+            I like doing it
+            it was easy'],
+        ];
+
+        $expected = [
+            'keyA' => ['This is fun
+            it was easy'],
+        ];
```
Comment (dickolsson) [inline tests/ThreeWayMergeTest.php:397]: I think this is wrong? Both local and remote version have added `I like doing it`, but it's not in the merged version? Does it have something to do with the fact that local doesn't have `PHP_EOL` at the end of it's line, but remote does. This would be why they aren't comparing equal?

File: src/ThreeWayMerge.php [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+                $this->linesAdded(
+                    $ancestor,
+                    $local,
+                    $remote,
+                    $count,
+                    $count_remote,
+                    $count_ancestor,
+                    $count_local
+                ) :
+                $this->linesRemovedOrModified($ancestor, $local, $remote, $count);
```
Comment (dickolsson) [inline src/ThreeWayMerge.php]: To improve readability I think sections like this one should be broken out into full `if` `else` statements, instead of the shorthand versions of it.

File: src/ThreeWayMerge.php [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+
+    protected function linesAdded(
+        array $ancestor,
+        array $local,
+        array $remote,
+        $count,
+        $count_remote,
+        $count_ancestor,
+        $count_local
+    ) {
```
Comment (dickolsson) [inline src/ThreeWayMerge.php:153]: In my opinion, breaking up the declaration of a method over multiple lines decreases the readability. I think PSR-2 states that this "MAY" be done, which means that it's not required. Can we try adding all function declarations on a single line to see if that improves the readability?

Comment (dickolsson) [inline src/ThreeWayMerge.php:153]: Or was this done to come in under the line length limit?

**Comment by dickolsson** [comment]
@sharprakeshverma I've taken a brief look, and in general it looks very good. Below is some general feedback:
- I might need to refresh my mind on the 3-way merge algorithm, but I left a few comments in the code where I think the merge logic is wrong
- The readability of the code is going to be crucial, especially because it's a lot of complicated logic. So let's do everything we can to break up logic to make things clear.
- This branch should not change `.travis.yml`, `build.xml` or `composer.json` in any way, since those changes are already in `master`. So please rebase or revert those changes.

File: tests/ThreeWayMergeTest.php (hunk 7 above)
Comment (rakesh-verma-16) [inline tests/ThreeWayMergeTest.php:364]: Start comparing them line by line from the very first line. The final merged version would carry only the changes made in any one of entities on a line and if two entities has modified the same line, it'd return a conflict error. As per recursive 3 way merge: The first line `This is not fun` is edited only on my local, so it'd detect the changes and make the change as per local in first line of merged version as well. Then comes the second line which says `I like doing it` has been modified(removed) in remote only and hence, it'd be removed from merged version as well. Similarly the third line `It was easy` has been modified(Added) in local node only and since these lines hasn't been modified in any other entities, it'd not return a conflict error.

Comment (rakesh-verma-16) [inline tests/ThreeWayMergeTest.php:364]: We expect `it was easy` here because according to 3 way merge, we don't find what is in 2 nodes, we look for what has been modified because the person who modifies it knows what he is doing. and if 2 person modifies different lines on same nodes, we will merge the changes in final version on their respective lines. and if they've modified the same lines, then it'd return a conflict which needs to be solved manually. Here, `it was easy` was added by the developer on the local system and original and remote haven't made any change on that line. This is why we expect that line. You can go through http://www.drdobbs.com/tools/three-way-merging-a-look-under-the-hood/240164902 to clear the idea again. Please let me know if it's still not correct

File: src/ThreeWayMerge.php (hunk 10 above)
Comment (rakesh-verma-16) [inline src/ThreeWayMerge.php:153]: To come under the line length limit.

File: tests/ThreeWayMergeTest.php (hunk 8 above)
Comment (rakesh-verma-16) [inline tests/ThreeWayMergeTest.php:397]: You are right for this line `I like doing it` in the testcase3() and testcase4(). It's a special edge case which doesn't follow normal rules. Added special condition for it now :+1:

File: tests/ThreeWayMergeTest.php (hunk 7 above)
Comment (dickolsson) [inline tests/ThreeWayMergeTest.php:364]: Where is the code for the special condition? I can't seem to find it...

File: src/ThreeWayMerge.php [hunk 11]
```diff
@@ -6,52 +6,316 @@
...
+     *
+     * @return mixed
+     * @throws Exception
+     */
+    protected function merge($x, $y, $z, $key)
+    {
+        // Convert the value into array.
+        $ancestor = (strpos($x, "\n") !== false ? explode("\n", $x) : array($x));
+        $local = (strpos($y, "\n") !== false ? explode("\n", $y) : array($y));
+        $remote = (strpos($z, "\n") !== false ? explode("\n", $z) : array($z));
```
Comment (dickolsson) [inline src/ThreeWayMerge.php]: Use the `PHP_EOL` instead of `\n`.

File: src/ThreeWayMerge.php [hunk 12]
```diff
@@ -6,52 +6,316 @@
...
+        // For example : addition in remote and removal in local node.
+        //
+        // Else get the value of number of lines in updated node.
+        // For example : Addition of 2 lines in local.
+        // Suppose: $count_ancestor = 2 and $count_remote = 2
+        // then, $count_local would be 4 now.
+        if ($count_ancestor != $count_local
+            && $count_local!= $count_remote
+            && $count_ancestor != $count_remote
+        ) {
```
Comment (dickolsson) [inline src/ThreeWayMerge.php:87]: When breaking up conditions over multiple lines I find it easier to read if you do this:

``` php
if (
  $count_ancestor != $count_local
  && $count_local!= $count_remote
  && $count_ancestor != $count_remote
) {
```

Don't you agree?

File: src/ThreeWayMerge.php [hunk 13]
```diff
@@ -6,52 +6,316 @@
...
+                    if (!isset($local[$key])) {
+                        unset($merged[$key]);
+                    }
+                } else {
+                    throw new Exception("A conflict has occured");
+                }
+            } elseif ($count_remote == $mincount
+                && ($count_ancestor == $maxcount
+                    || $count_local == $maxcount)) {
+                if ($local[$key] == $ancestor[$key]) {
```
Comment (dickolsson) [inline src/ThreeWayMerge.php:308]: For all conditions above that span multiple lines, let's do like this instead:

``` php
} elseif (
  $count_remote == $mincount
  && ($count_ancestor == $maxcount || $count_local == $maxcount)
) {
```

Notice above how I kept the condition within the `( )` on a single line. That'll be a lot more clear, since it still fits within the line limit.

File: src/ThreeWayMerge.php [hunk 14]
```diff
@@ -6,52 +6,316 @@
...
+
+        // First for loop compares all 3 nodes and returns updated node.
+        for ($key = 0; $key < $mincount; $key++) {
+            if ($ancestor[$key] == $local[$key]) {
+                $merged[$key] = $remote[$key];
+            } elseif ($ancestor[$key] == $remote[$key]
+                || $local[$key] == $remote[$key]) {
+                $merged[$key] = $local[$key];
+            } else {
+                throw new Exception("A conflict has occured");
```
Comment (dickolsson) [inline src/ThreeWayMerge.php]: Let's define and throw our own `ConflictException` class that extends the core `Exception` class.

File: src/ThreeWayMerge.php (hunk 12 above)
Comment (rakesh-verma-16) [inline src/ThreeWayMerge.php:87]: Yes but it'd throw an error saying "Every line in multi-line if statement must begin with a boolean operator"

Comment (rakesh-verma-16) [inline src/ThreeWayMerge.php:87]: I tried doing it but it seems like it Travis build won't pass. So I had to undo all the changes.

File: src/ThreeWayMerge.php (hunk 13 above)
Comment (rakesh-verma-16) [inline src/ThreeWayMerge.php:308]: Travis Build won't pass. Tried it.

File: src/ThreeWayMerge.php (hunk 14 above)
Comment (rakesh-verma-16) [inline src/ThreeWayMerge.php]: Done

File: src/ThreeWayMerge.php (hunk 11 above)
Comment (rakesh-verma-16) [inline src/ThreeWayMerge.php]: Done

File: tests/ThreeWayMergeTest.php (hunk 7 above)
Comment (rakesh-verma-16) [inline tests/ThreeWayMergeTest.php:364]: Please have a look at `line 282-289` in `ThreewayMerge.php`. This condition will handle the case where new lines has been added in both the remote and local and if they are equal, it'd merge them in revision else thrown an error.

File: src/ThreeWayMerge.php (hunk 11 above)
Comment (dickolsson) [inline src/ThreeWayMerge.php]: @sharprakeshverma It's only done when you have pushed the code ;)

File: src/ThreeWayMerge.php (hunk 12 above)
Comment (dickolsson) [inline src/ThreeWayMerge.php:87]: @sharprakeshverma Ok, fair enough then.

File: src/ThreeWayMerge.php [hunk 15]
```diff
@@ -1,57 +1,333 @@
...
+                && ($count_remote == $maxcount || $count_local == $maxcount)
+            ) {
+                if ($local[$key] == $remote[$key]) {
                     $merged[$key] = $local[$key];
+                } else {
+                    throw new ConflictException("A conflict has occured");
+                }
+            } elseif ($count_local == $mincount
+                && ($count_ancestor == $maxcount
+                    || $count_remote == $maxcount)
```
Comment (dickolsson) [inline src/ThreeWayMerge.php]: The condition within `( )` probably fits into one line here.

File: tests/ThreeWayMergeTest.php [hunk 16]
```diff
@@ -196,12 +196,648 @@ public function testConflict()
...
+                ],
+                2 => 'a little new something',
+            ]
+        ];
+        $merge = new ThreeWayMerge();
+        $result = $merge->performMerge($original, $local, $remote);
+        $this->assertEquals($expected, $result);
+    }
+    // Conflict in KeyA=>0=>KeyB.
+    public function testMultilineConflict1()
```
Comment (dickolsson) [inline tests/ThreeWayMergeTest.php:576]: Let's add proper documentation block to each test method, briefly describing the test. For example like this:

``` php
/**
 * Test that assert conflicts when a key are modified in both versions.
 */
```

File: tests/ThreeWayMergeTest.php [hunk 17]
```diff
@@ -196,12 +196,648 @@ public function testConflict()
...
+
+        $expected = [
+            'keyA' => ['This is fun
+            it was easy'],
+        ];
+        $merge = new ThreeWayMerge();
+        $result = $merge->performMerge($original, $local, $remote);
+        $this->assertEquals($expected, $result);
+    }
+    public function testMultilineRecursiveAdditionRemovalModification()
```
Comment (dickolsson) [inline tests/ThreeWayMergeTest.php]: There should be a blank line between methods.

File: tests/ThreeWayMergeTest.php (hunk 7 above)
Comment (dickolsson) [inline tests/ThreeWayMergeTest.php:364]: Ok, I see.

File: src/ThreeWayMerge.php [hunk 18]
```diff
@@ -1,57 +1,333 @@
...
+                        $remote[$key],
+                        $key
+                    );
+                } elseif (array_key_exists($key, $local)) {
+                    if ($ancestor[$key] != $local[$key]) {
+                        throw new ConflictException("A conflict has occured");
+                    }
+                } elseif (array_key_exists($key, $remote)) {
+                    if ($ancestor[$key] != $remote[$key]) {
+                        throw new ConflictException("A conflict has occured");
```
Comment (dickolsson) [inline src/ThreeWayMerge.php:44]: It would be kind of cool if we added some basic info about the conflict in the message. Maybe something like:

``` php
throw new ConflictException("Merge conflict in $key");
```

And grammatically speking, the conflict doesn't "occur" (i.e. it doesn't happen as we parse it), it's technically already there. So let's use something like "Merge conflict in $key".

File: src/ThreeWayMerge.php (hunk 13 above)
Comment (dickolsson) [inline src/ThreeWayMerge.php:308]: Ok.

File: src/ThreeWayMerge.php (hunk 18 above)
Comment (rakesh-verma-16) [inline src/ThreeWayMerge.php:44]: Well, I'm afraid we might not be able to do it as easily as `throw new ConflictException("Merge conflict in $key);` as our keys here are the number of lines in the value and not the actual keys. what we can do is make every method to return a particular value if conflict arises. and in the main function of ours, we can check. If a conflict arises, then we can return this exception otherwise we can just update the revision. What do you think?

Comment (rakesh-verma-16) [inline src/ThreeWayMerge.php:44]: Even then it would probably return `"Merge Conflict in keyB"` without giving any prior knowledge about it's parent key. For ex: 1=>keyB or 2=>keyB?

File: tests/ThreeWayMergeTest.php [hunk 19]
```diff
@@ -196,12 +196,783 @@ public function testConflict()
...
+        } catch (ConflictException $e) {
             $this->assertTrue(true);
         }
     }
+
+    /**
+     * Test for multiline modification
+     * in same number of lines.
+     *
+     * @throws ConflictException
```
Comment (dickolsson) [inline tests/ThreeWayMergeTest.php]: None of the test methods throws `ConflictException`. In the cases we do test conflicts we catch the exception, so it's not thrown in that sense. So let's remove this line for all test methods.

File: src/ThreeWayMerge.php (hunk 18 above)
Comment (dickolsson) [inline src/ThreeWayMerge.php:44]: I don't think it's too important to give a full clue of the parent keys. As long as we give _some_ clue it's good I think. In multiline merge we can say "Merge conflict on line $key" and in other situations we can just say "Merge conflict in key $key".

That's better than a static message.

File: tests/ThreeWayMergeTest.php [hunk 20]
```diff
@@ -196,12 +196,783 @@ public function testConflict()
...
+     * Test to assert the Addition, removal and
+     * modification in the values all at the same time.
+     *
+     * @throws ConflictException
+     */
+    public function testMultilineAdditionRemovalModification3()
+    {
+        $original = [
+                'keyA' => ['This is not fun'],
+            ];
```
Comment (dickolsson) [inline tests/ThreeWayMergeTest.php]: The indentations here look a bit strange.

File: README.md [hunk 21]
```diff
@@ -1 +1,108 @@
-PHP Library to perform 3 way merge on associative arrays.
+# Relaxedws/merge [![Build Status](https://travis-ci.org/relaxedws/merge.svg?branch=master)](https://travis-ci.org/relaxedws/merge)
+
+A Library to perform recursive 3 way merge algorithm
```
Comment (dickolsson) [inline README.md]: Should be `A library ...`

File: tests/ThreeWayMergeTest.php (hunk 20 above)
Comment (rakesh-verma-16) [inline tests/ThreeWayMergeTest.php]: I wonder why no error was thrown.

File: README.md [hunk 22]
```diff
@@ -1 +1,108 @@
...
+on associative arrays, written in PHP.
+
+## Insight
+
+This library is built to perform a recursive 3 way merge algorithm. It takes 3 parameters which are arrays representing base entity, local entity and remote entity. It compares each of these entities with other entities line by line.
+If only one out of remote or local is updated out of these 3, the final revision will have all the unchanged data in it along with the update data from the update entity(Either remote or local). If more than one entity is updated on the same line, it'd throw a `ConflictException`.
```
Comment (dickolsson) [inline README.md]: Missing space between word and parenthesis  `...entity(Either...`.

File: README.md (hunk 21 above)
Comment (rakesh-verma-16) [inline README.md]: It is already `A library to perform`. Can you please explain where you are pointing?

Comment (dickolsson) [inline README.md]: Also, we should always use the correct punctuation for 3-way merge algorithm, i.e. use "-" between "3" and "way".

File: README.md (hunk 22 above)
Comment (dickolsson) [inline README.md]: Let's also replace "entity" with "array" everywhere in the text, since this is a generic PHP library and the arrays could be anything.

File: README.md [hunk 23]
```diff
@@ -1 +1,108 @@
...
+      "Relaxed\\Merge\\": "src/"
+    }
+  },
+  "autoload-dev": {
+    "psr-4": {
+      "Relaxed\\Merge\\Test\\": "tests/"
+    }
+  }
+}
+````
```
Comment (dickolsson) [inline README.md:24]: Using our own composer file doesn't make sense here. It's better to show how another library would add a dependency on this library. For example:

``` JSON
{
  "name": "myorg/mylib",
  "description": "A library depending on 3-way merge",
  "require": {
    "relaxedws/merge": "dev-master",
  }
}
```

File: README.md [hunk 24]
```diff
@@ -1 +1,108 @@
...
+        1 => [
+            'keyB' => 'This is milk',
+            'keyC' => 'This is changed because of remote',
+        ],
+        2 => 'a little sugar',
+    ]
+];
+
+$multiline = new ThreeWayMerge();
+$new_arr = $multiline->performMerge($original, $local, $remote);
```
Comment (dickolsson) [inline README.md]: I don't think it makes sense to name the merge object "$multiline". Let's just use "$merge".

File: README.md [hunk 25]
```diff
@@ -1 +1,108 @@
...
+Though it's working well for almost all possible scenarios but we all know there are bugs
+in every code. Although we keep updating and testing the library regulary, but still if
+you encounter any bugs, please feel free to drop us a comment with as much explaination as
+possible to reproduce the errors or you can even patch that.
+
+Other than directly working with the code, you can also help us with better documentation, fixing typo errors
+or adding more content to the `readme` file.
+
+We are open to feedback, suggestions and questions.
+We are always present at ``#drupal8-ports` on `irc.freenode.net` .
```
Comment (dickolsson) [inline README.md]: Let's not mix in Drupal 8 here. If people wants to contribute to this library we should encourrage them to submit pull requests. And if they have questions they should file Github issues for this project. That's better, because then other people can search among these issues and maybe find answers to their own questions etc :)

File: README.md [hunk 26]
```diff
@@ -1 +1,108 @@
...
+$new_arr = $multiline->performMerge($original, $local, $remote);
+````
+
+## Contributing
+
+We welcome all developers to come forward for use/test of this library.
+Though it's working well for almost all possible scenarios but we all know there are bugs
+in every code. Although we keep updating and testing the library regulary, but still if
+you encounter any bugs, please feel free to drop us a comment with as much explaination as
+possible to reproduce the errors or you can even patch that.
```
Comment (dickolsson) [inline README.md]: Here's a shorter, less rambling version ;)

```
We welcome anyone to use, test, or contribute back to this project.
We have extensive test coverage, but as we all know there's always bugs in software.
Please file issues or pull requests with your comments or suggestions.
```

File: src/ThreeWayMerge.php (hunk 18 above)
Comment (rakesh-verma-16) [inline src/ThreeWayMerge.php:44]: On it :D
